Infra/masked input to ts

ethanshar opened the pull request:
## Description
Remove usage of legacy TextField from MaskedInput and migrate component to TS and function component

## Changelog
Refactor MaskedInput component, remove legacy TextField usage

M-i-k-e-l reviewed (commented):
Consider moving the screen to TS as well.
Also, the screen did not work as expected for me, typing more than one char in the first example did not always work.

ethanshar commented on src/components/maskedInput/index.tsx:
```diff
@@ -1,90 +1,111 @@
+
+  useEffect(() => {
+    keyboardDidHideListener.current = Keyboard.addListener('keyboardDidHide', () => {
+      if (inputRef.current?.isFocused?.()) {
+        inputRef.current?.blur?.();
```
Why not? It's safer, no?

ethanshar commented on src/components/maskedInput/index.tsx:
```diff
@@ -1,90 +1,111 @@
+  useImperativeHandle(ref, () => {
+    return {
+      isFocused: () => inputRef.current?.isFocused(),
+      focus: () => inputRef.current?.focus(),
```
I'm not sure what you mean..

ethanshar commented on src/components/maskedInput/index.tsx:
```diff
@@ -1,90 +1,111 @@
 import Text from '../text';
 import TouchableOpacity from '../touchableOpacity';
 
+export interface MaskedInputProps extends TextInputProps {
```
I think that extending TextField, the old or the new is creating unnecessary complexity for this component. 
But I see your point, so I created a middleware migrator file and added a migrate prop, so users can move to the new implementation when they are ready. 

For usages in One App, no one uses any of the TextField features

M-i-k-e-l reviewed (commented):
I could not comment on older conversations so I had to re-create them :/

M-i-k-e-l commented:
> Consider moving the screen to TS as well. 

I think you might have missed this.

> Also, the screen did not work as expected for me, typing more than one char in the first example did not always work.

I think that using the computer's keyboard is the issue, using the simulator \ Android works.

ethanshar commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,119 @@
+  useEffect(() => {
+    keyboardDidHideListener.current = Keyboard.addListener('keyboardDidHide', () => {
+      if (inputRef.current?.isFocused?.()) {
+        inputRef.current?.blur?.();
```
Fixed

ethanshar commented on src/index.ts:
```diff
@@ -86,7 +86,6 @@ export {default as Icon, IconProps} from './components/icon';
 export {default as ListItem, ListItemProps} from './components/listItem';
 export {default as LoaderScreen, LoaderScreenProps} from './components/loaderScreen';
-// @ts-expect-error
 export {default as MaskedInput} from './components/maskedInput';
```
Exported

ethanshar commented:
@M-i-k-e-l 
I fixed the PR comments
I also changed a little the "new" implementation after playing with in the private components. 
First, I turned it into an uncontrolled component because of how the `formatter` (used to be called `transformer`) used to work. So I also added an `initialValue` prop

M-i-k-e-l commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,129 @@
+  [onChangeText, formatter]);
+
+  const focus = useCallback(() => {
+    inputRef.current?.focus?.();
```
```suggestion
    inputRef.current?.focus();
```

M-i-k-e-l commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,129 @@
+import Text from '../text';
+import TouchableOpacity from '../touchableOpacity';
+
+export interface MaskedInputProps extends TextInputProps {
```
```suggestion
export interface MaskedInputProps extends Omit<TextInputProps, 'value'> {
```

M-i-k-e-l commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,129 @@
+ * @gif: https://camo.githubusercontent.com/61eedb65e968845d5eac713dcd21a69691571fb1/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f4b5a5a7446666f486f454b334b2f67697068792e676966
+ * @example: https://github.com/wix/react-native-ui-lib/blob/master/demo/src/screens/componentScreens/MaskedInputScreen.js
+ */
+function MaskedInput(props: MaskedInputProps, ref: ForwardedRef<any>) {
```
I've written this in the main PR area, they should make that more noticeable somehow...
> Consider moving the screen to TS as well.

ethanshar replied:
Definitely. Fixed and also add a new cool example (:

M-i-k-e-l commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,129 @@
+ */
+function MaskedInput(props: MaskedInputProps, ref: ForwardedRef<any>) {
+  const {initialValue, formatter = _.identity, containerStyle, renderMaskedText, onChangeText, ...others} = props;
+  const [value, setValue] = useState(initialValue);
```
Do we want to format the `initialValue` here?

M-i-k-e-l commented on src/components/maskedInput/new.tsx:
```diff
@@ -0,0 +1,129 @@
+  });
+
+  useEffect(() => {
+    if (initialValue !== value) {
```
Do we want to format the `initialValue` here?

ethanshar replied:
I guess the same answer as the other comment

M-i-k-e-l commented on demo/src/screens/componentScreens/MaskedInputScreen.tsx:
```diff
@@ -56,6 +54,20 @@ export default class MaskedInputScreen extends Component {
+      <View row centerH>
+        {_.times(4, i => {
+          return (
+            <View marginR-s3 center style={styles.pinCodeSquare}>
```
Missing `key`

ethanshar replied:
right, fixed!

M-i-k-e-l commented on demo/src/screens/componentScreens/MaskedInputScreen.tsx:
```diff
@@ -85,11 +97,24 @@ export default class MaskedInputScreen extends Component {
+          <MaskedInput
+            migrate
+            maxLength={4}
+            ref={this.priceInput}
```
You should create a dedicated `ref`, currently only one of the two users (price \ pin) are deleted.

ethanshar replied:
Fixed as well (:
